### PR TITLE
Fix a regression where libthrift 0.9.1 is not compatible with `thrift0.9`

### LINKED ISSUE #4424: Make `armeria-thrift0.9` module compatible with libthrift 0.9.1
`armeria-thrift0.9` uses libthrift 0.9.3.1 to ensure compatibility with Thrift 0.9.x.
But I was told by @mauhiz that `armeria-thrift0.9` no longer supports Thrift 0.9.1.

#4226 introduced a way to limit max message length which uses the following constructor added in thrift [0.9.2](https://github.com/apache/thrift/commit/cd9dd515bb2bc49459af1114c40f020c96e87b41)
https://github.com/line/armeria/blob/a5ec2c2a0e81f20ab215f06023488b299acd2312/thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java#L125

We need some reflection tweaks to find an available constructor to support Thrift 0.9.1.

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

`thrift0.9` module is made to support Thrift 0.9.x versions. However, it is not compatible with 0.9.1 version with Armeria 1.17.0 to 1.19.0. The regresion is caused by #4226 which used an API that was introduced in Thrift 0.9.2

Modifications:

- Add a testing module that ensures compatibility with Thrift 0.9.1
- Generated files by Thrift compiler 0.9.1 are added because we have already upgraded Thrift compiler version to 0.9.3-1 for `thrift0.9` module but the generated files by 0.9.3-1 compiler is not compatible with libthrift 0.9.1
- Exclude some test files which is not compatible with libthrift 0.9.1
- Fix `ThriftProtocolFactories` to use the non-arguments constructor for libthrift 0.9.1

Result:

- Fixed a regression where `thrift0.9` is incompatible with Thrift 0.9.1 since 1.17.0
- Fixes #4424

### CODE REVIEW

File: it/thrift0.9.1/src/test/java/java/com/linecorp/armeria/package-info.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+/**
+ * All sub-packages of this package contain generated files by Thrift Compiler 0.9.1.
+ * The generated code is intentionally added because Thrift compiler 0.9.3 is not compatible with
+ * libthrift 0.9.1.
```
Comment (ikhoon) [inline it/thrift0.9.1/src/test/java/java/com/linecorp/armeria/package-info.java:20]: Note for reviewers.
Only this file and `it/thrift0.9.1/build.gradle` are created by hand. 😆

**Review by minwoox — COMMENTED** [review]
Besides the compiler problem we had talked about offline, just one small nit.
Thanks! 😄

File: thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java [hunk 2]
```diff
@@ -35,6 +37,21 @@
...
+    private static final boolean IS_THRIFT_091;
+
+    static {
+        boolean isThrift091 = false;
+        try {
+            final Constructor<TBinaryProtocol.Factory> ignored =
+                    TBinaryProtocol.Factory.class.getConstructor(boolean.class, boolean.class,
+                                                                 long.class, long.class);
+        } catch (NoSuchMethodException e) {
+            e.printStackTrace();
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java:49]: 😱

[the remark this answers is not in the document]
Comment (ikhoon) [reply thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java]: Remove it.

File: thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java [hunk 3]
```diff
@@ -120,7 +136,21 @@ public static TProtocolFactory binary(int maxStringLength, int maxContainerLengt
         final int maxStringLength0 = maxStringLength == 0 ? -1 : maxStringLength;
         final int maxContainerLength0 = maxContainerLength == 0 ? -1 : maxContainerLength;
 
-        // Thrift 0.9.x does not a constructor taking only maxStringLength and maxContainerLength.
+        if (IS_THRIFT_091) {
+            // Thrift 0.9.1 does not support a message length limit.
```
Comment (jrhee17) [inline thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java:140]: I'm wondering if it is better to validate this at `ThriftSerializationFormats`.

At the very least, I think we may want to document this behavior at the following location:
https://github.com/line/armeria/blob/1179e84f06958115d0d032d605f53f8648d3ada2/thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftSerializationFormats.java#L112

**Review by jrhee17 — COMMENTED** [review]
Looks good overall! Left a minor comment 🙏

File: thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java (hunk 3 above)
Comment (ikhoon) [reply thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java:140]: Nice suggestion. A validation close to `.protocolFactory()` would be useful. I will add an additional validation to `protocolFactory(SerializationFormat,int,int)` which will raise IllegalArgumentException if length limits are set and the Thrift version is 0.9.1.

Comment (ikhoon) [reply thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java:140]: I just realized that a max request or response length is passed to `maxStringLength` to create the default one.
We need some refactoring to fix the default value depending on Thrift versions.
https://github.com/line/armeria/blob/a5ec2c2a0e81f20ab215f06023488b299acd2312/thrift0.13/src/main/java/com/linecorp/armeria/internal/client/thrift/THttpClientDelegate.java#L100-L106
I'm not sure we need such a change for this legacy version. Let me leave a warning log and update Javadoc's description.

File: it/thrift0.9.1/src/test/thrift/RpcDebugTest.thrift [hunk 4]
```diff
@@ -0,0 +1,21 @@
+namespace java com.linecorp.armeria.common.thrift.text
```
Comment (minwoox) [inline it/thrift0.9.1/src/test/thrift/RpcDebugTest.thrift:1]: Can't we also copy these Thrift IDLs?

Comment (minwoox) [reply it/thrift0.9.1/src/test/thrift/RpcDebugTest.thrift:1]: Oops, never mind the comment. 😉

File: thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java (hunk 3 above)
Comment (minwoox) [reply thrift0.13/src/main/java/com/linecorp/armeria/common/thrift/ThriftProtocolFactories.java:140]: I'm fine with the warning log.

**Review by minwoox — APPROVED** [review]
Thanks! 👍 👍 👍

**Review by jrhee17 — APPROVED** [review]
Thanks @ikhoon ! 🙇 👍 🙇

**Comment by minwoox** [comment]
🎉 🎉 🎉

**Pull request merged**